### ISSUE TITLE
Unauthorized access to this file for pasted images

### ISSUE THREAD

**Issue by Akuket** [glpi-project/glpi#5457]
Hello,

**Describe the bug**

One user reported to us that he can't view a file pasted in one ticket while everyone should be able to see this one. After some investigations, it appears that only administrators can see it. 
The entry point of that problem could be the method canViewFile from the document.class.php file.

**Page(s) URL**

/front/ticket.form.php and /front/document.send.php

**To reproduce**

In a ticket: copy-paste an image in a rich text field (description/follow-up/task)

**Expected behavior**

Everyone that can see the ticket can see the image too.

**Screenshots**

![image](https://user-images.githubusercontent.com/32392661/53255635-4f196680-36c6-11e9-9e0a-6d64d14bed5b.png)

![image](https://user-images.githubusercontent.com/32392661/53255677-622c3680-36c6-11e9-8392-b733a6a0c1f8.png)

**Your GLPI setup (you can find it in Setup > General menu, System tab)**

Tested with GLPI 9.3.3 and GLPI 9.4.0

**Additional context**

We had already encountered this problem before. It had been mentioned in that issue https://github.com/glpi-project/glpi/issues/4236 and then corrected. It seems to reappear. Can this be the cause of a recent change or the problem is not fully resolved?

Regards

**Comment by cedric-anne** [comment]
Hi,

The problem, for tickets, is only on solutions. Looks like #4236 was only fixing followups.

Can you test #5472 ?

Regards

**Comment by Akuket** [comment]
Hi,

I replaced my files by the 3 in your pull request, but the problem still occurs.

Regards

**Comment by cedric-anne** [comment]
Hi,

Sorry, my question was not enough precise.
The fix should work but only for solutions submitted once it is applied. For existing data, I have to add a migration script.

Was your test made on existing solutions or on new ones ?

Regards

**Comment by Akuket** [comment]
It was on existing solutions

**Comment by trasher** [comment]
Then, could you please test if you reproduce on a new one with proposed fix?

I guess it is OK, we can now think about the migration part.

**Comment by Akuket** [comment]
We have some solutions with pictures pasted inside. So I don't know if my test will be conclusive.

If you write a migration I could test it with the broken solution.

**Comment by Akuket** [comment]
Hello,
The problem still occurs with the glpi 9.4.1 version.
![image](https://user-images.githubusercontent.com/32392661/55152255-434f1300-5150-11e9-811b-6634a62eaeeb.png)

Regards

**Comment by cedric-anne** [comment]
Merged Pr in 9.4.1 contains a migration (see https://github.com/glpi-project/glpi/commit/de4ba9b9d24f3b0aef20e74611933d75d5236248#diff-05240cc13481ecf278461feed763ec72R55 ).

Can you check the URL of the image ? If it starts with `//`, then it may be related to #5619.

Regards

**Comment by Akuket** [comment]
No, the url doesn't start with //
And I can see the image with an admin account, but not with otherwise.

**Comment by cedric-anne** [comment]
Hi,

What is the URL of the image and in which context is it displayed ?

Regards

**Comment by Akuket** [comment]
Hi,
The url is https://domain/front/document.send.php?docid=391
With an administrator account, I can see the image. As an observer or being assigned to the ticket, it's impossible.
Regards

**Comment by cedric-anne** [comment]
I made a test and cannot reproduce. I tried to run the migration and it works.

If it is on a solution made prior to updating in 9.4.1, and if you had use a nightly 9.4.x version in production (bad idea!), maybe the migration did not run again while updating. In this case, you can try to use `php bin/console db:update -n -f` but DO A BACKUP FIRST.

If you find a way to reproduce this on a new solution, please open a new ticket and describe the steps made to reproduce.

Regards

**Comment by Akuket** [comment]
It wasn't a nightly 9.4.x version. I tried your command and nothing change for this bug.
Currently, I don't know how to reproduce this one.

Regards

**Comment by Akuket** [comment]
We move forward. I succeeded to identify one part of the problem.
Some of our documents don't have the tickets_id field filled in database and the page /front/ticket.form.php doesn't provide the tickets_id in the url of the link as you can see below

```html
<p>
  <a href="/front/document.send.php?docid=391" target="_blank">
    <img alt="06a5c38e-a5064f86-5c6d833aa54cb2.93194692" width="484.00000000000006" src="/front/document.send.php?docid=391">
  </a>
</p>
```

If you can change the html above to that 
```html
<p>
  <a href="/front/document.send.php?docid=391&tickets_id=2158" target="_blank">
    <img alt="06a5c38e-a5064f86-5c6d833aa54cb2.93194692" width="484.00000000000006" src="/front/document.send.php?docid=391&tickets_id=2158">
  </a>
</p>
```

and always adding the tickets_id in glpi_documents items, the problem will be solved.

Do you think that would be feasible?

Regards

**Comment by cedric-anne** [comment]
Hi,

The `tickets_id` field from `glpi_documents` should not be used anymore in timeline display in 9.4.x, except if I missed something.

So, I ask again, in which context this image is displayed ? If it is on ticket timeline, can you give me the exact data existing in database for the element (from `glpi_itilfollowups` if in a followup, from `glpi_tickettasks` in in a task, from `glpi_itilsolutions` if in a solution ?

Regards

**Comment by Akuket** [comment]
Hi,

This image is displayed on the solution, but I can't find her in the glpi_itilsolutions table. The docid of this image is 391. To find this one, I must use this SQL rule?  select * from glpi_itilsolutions where id=391;

Regards

**Comment by cedric-anne** [comment]
Hi,

`SELECT * FROM `glpi_itilsolutions` WHERE `itemtype` = 'Ticket' AND `items_id` = '2158'`

Regards

**Comment by Akuket** [comment]
id: 1034

itemtype: Ticket

solutiontypes_id: 0

solutiontype_name: NULL

content: &lt;p&gt;&lt;a href=&quot;/front/document.send.php?docid=391&quot; target=&quot;_blank&quot; &gt;&lt;img alt=&quot;06a5c38e-a5064f86-5c6d833aa54cb2.93194692&quot; width=&quot;484.00000000000006&quot; src=&quot;/front/document.send.php?docid=391&quot; /&gt;&lt;/a&gt;&lt;/p&gt;&lt;p&gt;&nbsp;&lt;/p&gt;

date_creation: 20/02/2019  17:41:31

date_mod: 20/02/2019  17:41:31

date_approval: 22/02/2019  15:51:38

users_id: 92

user_name: NULL

users_id_editor: 0

users_id_approval: 2

user_name_approval: NULL

status: 4

itilfollowups_id: NULL

**Comment by cedric-anne** [comment]
Please give me the mysql export of this line,  as I need to be sure to have the raw data that is processed by migration.

**Comment by Akuket** [comment]
+------+----------+----------+------------------+-------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------+---------------------+---------------------+----------+-----------+-----------------+-------------------+--------------------+--------+------------------+
| id   | itemtype | items_id | solutiontypes_id | solutiontype_name | content                                                                                                                                                                                                                                                                                                       | date_creation       | date_mod            | date_approval       | users_id | user_name | users_id_editor | users_id_approval | user_name_approval | status | itilfollowups_id |
+------+----------+----------+------------------+-------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------+---------------------+---------------------+----------+-----------+-----------------+-------------------+--------------------+--------+------------------+
| 1034 | Ticket   |     2158 |                0 | NULL              | &lt;p&gt;&lt;a href=&quot;/front/document.send.php?docid=391&quot; target=&quot;_blank&quot; &gt;&lt;img alt=&quot;06a5c38e-a5064f86-5c6d833aa54cb2.93194692&quot; width=&quot;484.00000000000006&quot; src=&quot;/front/document.send.php?docid=391&quot; /&gt;&lt;/a&gt;&lt;/p&gt;&lt;p&gt;&nbsp;&lt;/p&gt; | 2019-02-20 17:41:31 | 2019-02-20 17:41:31 | 2019-02-22 15:51:38 |       92 | NULL      |               0 |                 2 | NULL               |      4 |             NULL |
+------+----------+----------+------------------+-------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------+---------------------+---------------------+----------+-----------+-----------------+-------------------+--------------------+--------+------------------+

**Comment by cedric-anne** [comment]
I have no explaination.

Your solution has been submitted prior to 9.4.1 version according to the `date_creation`.
If I put this data in a 9.4.0 database and do the update process to 9.4.1, the tickets_id is added on the URL and problem is solved.
If I put this data in a 9.4.1 database and force replay of last migration, problem is solved too.

So I cannot help anymore.

**Comment by Akuket** [comment]
Hello

Can you send me the new MySQL export of this line and the process you executed to have this result, please?
Thanks

Regards

**Comment by cedric-anne** [comment]
Hi,

As I said previously, I forced execution of the last migration script by using `php bin/console db:update -n -f`. 
It will just add `&tickets_id=XXX` in URLs where it is missing, as you noticed in https://github.com/glpi-project/glpi/issues/5457#issuecomment-478070491.

Regards

**Comment by Akuket** [comment]
I retry your command and nothing change for me :(

**Comment by Akuket** [comment]
I found ! In your migration you have this regex 
` $missing_param_pattern = '(document\.send\.php\?docid=\d+)(' . implode('|', $quotes_possible_exp) . ')';`
\d is not a valid set for our mysql. Please change it to [0-9].
Thanks

**Comment by cedric-anne** [comment]
Which database server do you use, and which version ?

**Comment by Akuket** [comment]
mysql server 5.7

**Comment by cedric-anne** [comment]
Hi,

I use MariaDB, which supports `\d` for digits, that's why I did not find the problem.
This will be fixed on 9.4.2 by #5750 .

Regards

**Comment by Akuket** [comment]
Thanks
